clustermesh-apiserver/main.go

```diff
@@ -365,202 +334,155 @@ func (n *nodeStub) GetKeyName() string {
 }
 
-func synchronizeNodes(clientset k8sClient.Clientset) {
-	_, ciliumNodeInformer := informer.NewInformer(
-		cache.NewListWatchFromClient(clientset.CiliumV2().RESTClient(),
-			"ciliumnodes", k8sv1.NamespaceAll, fields.Everything()),
-		&ciliumv2.CiliumNode{},
-		0,
-		cache.ResourceEventHandlerFuncs{
-			AddFunc: updateNode,
-			UpdateFunc: func(_, newObj interface{}) {
-				updateNode(newObj)
-			},
-			DeleteFunc: func(obj interface{}) {
-				deletedObj, ok := obj.(cache.DeletedFinalStateUnknown)
-				if ok {
-					deleteNode(deletedObj.Obj)
-				} else {
-					deleteNode(obj)
-				}
-			},
-		},
-		k8s.ConvertToCiliumNode,
-	)
+type ipmap map[string]struct{}
 
-	go ciliumNodeInformer.Run(wait.NeverStop)
+type endpointSynchronizer struct {
+	store store.SyncStore
+	cache map[string]ipmap
 }
 
-func updateEndpoint(oldEp, newEp *types.CiliumEndpoint) {
-	var ipsAdded []string
-	if n := newEp.Networking; n != nil {
+func newEndpointSynchronizer(ctx context.Context) synchronizer {
+	endpointsStore := store.NewWorkqueueSyncStore(kvstore.Client(), path.Join(ipcache.IPIdentitiesPath, ipcache.DefaultAddressSpace),
+		store.WSSWithSourceClusterName(cfg.LocalClusterName()))
+	go endpointsStore.Run(ctx)
+
+	return &endpointSynchronizer{
+		store: endpointsStore,
+		cache: make(map[string]ipmap),
+	}
+}
+
+func (es *endpointSynchronizer) upsert(ctx context.Context, key resource.Key, obj runtime.Object) error {
+	endpoint := obj.(*types.CiliumEndpoint)
+	ips := make(ipmap)
+	stale := es.cache[key.String()]
+
+	if n := endpoint.Networking; n != nil {
 		for _, address := range n.Addressing {
 			for _, ip := range []string{address.IPV4, address.IPV6} {
 				if ip == "" {
 					continue
 				}
 
-				keyPath := path.Join(ipcache.IPIdentitiesPath, ipcache.DefaultAddressSpace, ip)
+				scopedLog := log.WithFields(logrus.Fields{logfields.Endpoint: key.String(), logfields.IPAddr: ip})
 				entry := identity.IPIdentityPair{
 					IP:           net.ParseIP(ip),
-					Metadata:     "",
 					HostIP:       net.ParseIP(n.NodeIP),
-					K8sNamespace: newEp.Namespace,
-					K8sPodName:   newEp.Name,
+					K8sNamespace: endpoint.Namespace,
+					K8sPodName:   endpoint.Name,
 				}
 
-				if newEp.Identity != nil {
-					entry.ID = identity.NumericIdentity(newEp.Identity.ID)
+				if endpoint.Identity != nil {
+					entry.ID = identity.NumericIdentity(endpoint.Identity.ID)
 				}
 
-				if newEp.Encryption != nil {
-					entry.Key = uint8(newEp.Encryption.Key)
+				if endpoint.Encryption != nil {
+					entry.Key = uint8(endpoint.Encryption.Key)
 				}
 
-				marshaledEntry, err := json.Marshal(entry)
-				if err != nil {
-					log.WithError(err).Warningf("Unable to JSON marshal entry %#v", entry)
+				scopedLog.Info("Upserting endpoint in etcd")
+				if err := es.store.UpsertKey(ctx, &entry); err != nil {
+					// The only errors surfaced by WorkqueueSyncStore are the unrecoverable ones.
+					scopedLog.WithError(err).Warning("Unable to upsert endpoint in etcd")
```

Comment:
Shouldn't we also delete ip from stale in case of an errorhere?

Reply:
I'm not sure about that. With the current implementation, a possibly existing entry for that IP would be removed as considered stale. I personally feel that this is correct as behavior, given that we should have replaced it with a different one.

Reply:
ok, I guess it doesn't matter too much - now we are getting error here only when encoding fails.